File: 0123. Best Time to Buy and Sell Stock III/123.best-time-to-buy-and-sell-stock-iii.java
//You are given an array prices where prices[i] is the price of a given stock
//on the iᵗʰ day.
//
// Find the maximum profit you can achieve. You may complete at most two
//transactions.
//
// Note: You may not engage in multiple transactions simultaneously (i.e., you
//must sell the stock before you buy again).
//
//
// Example 1:
//
//
//Input: prices = [3,3,5,0,0,3,1,4]
//Output: 6
//Explanation: Buy on day 4 (price = 0) and sell on day 6 (price = 3), profit =
//3-0 = 3.
//Then buy on day 7 (price = 1) and sell on day 8 (price = 4), profit = 4-1 = 3.
//
//
// Example 2:
//
//
//Input: prices = [1,2,3,4,5]
//Output: 4
//Explanation: Buy on day 1 (price = 1) and sell on day 5 (price = 5), profit =
//5-1 = 4.
//Note that you cannot buy on day 1, buy on day 2 and sell them later, as you
//are engaging multiple transactions at the same time. You must sell before buying
//again.
//
//
// Example 3:
//
//
//Input: prices = [7,6,4,3,1]
//Output: 0
//Explanation: In this case, no transaction is done, i.e. max profit = 0.
//
//
//
// Constraints:
//
//
// 1 <= prices.length <= 10⁵
// 0 <= prices[i] <= 10⁵
//
//
// Related Topics Array Dynamic Programming 👍 9986 👎 203


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int K = 2;
        int[][][] dp = new int[n][K][2];
        for (int i = 0; i < n; i++) {
            for (int k = 0; k < K; k++) {
                if (i == 0) {
                    dp[i][k][0] = 0;
                    dp[i][k][1] = -prices[i];
                    continue;
                }
                dp[i][k][0] = Math.max(dp[i - 1][k][0], dp[i - 1][k][1] + prices[i]);
                if (k - 1 < 0) {
                    dp[i][k][1] = Math.max(dp[i - 1][k][1], -prices[i]);
                } else {
                    dp[i][k][1] = Math.max(dp[i - 1][k][1], dp[i - 1][k - 1][0] - prices[i]);
                }
            }
        }
        return dp[n - 1][K - 1][0];
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 0123. Best Time to Buy and Sell Stock III/123.best-time-to-buy-and-sell-stock-iii.go
package leetcode

/*
 * @lc app=leetcode id=123 lang=golang
 *
 * [123] Best Time to Buy and Sell Stock III
 */

// @lc code=start
func maxProfit(prices []int) int {
	// dp[i][k][0 or 1]
	// K: 允許買賣次數
	// n: 交易天數
	// 0: 代表不持有股票，1: 代表持有股票
	// 總計 n * K * 2 總狀態
	// 狀態轉移方程式
	// for 0 <= i < n:
	//     for 1 <= k <= K:
	//         for s in {0, 1}:
	//             dp[i][k][s] = max(buy, sell, rest)

	// 由於這題只能可以交易2次，所以 K 為 2
	n := len(prices)
	K := 2
	dp := make([][][]int, n)
	for i := range dp {
		dp[i] = make([][]int, K)
		for j := range dp[i] {
			dp[i][j] = make([]int, 2)
		}
	}

	for i := 0; i < n; i++ {
		for k := 0; k < K; k++ {
			if i == 0 {
				// 第一天不持有股票的情況，肯定沒有獲利
				dp[i][k][0] = 0
				// 第一天交易持有股票，代表為成本
				dp[i][k][1] = -prices[i]
				continue
			}
			// 第 i 天，交易累計 k 次，手上不持有股票
			// 1. 第 i - 1 天，交易累計 k 次，一樣沒持有，持續觀望
			// 2. 第 i - 1 天，交易累計 k 次，持有現價賣出
			dp[i][k][0] = max(dp[i-1][k][0], dp[i-1][k][1]+prices[i])
			// 第 i 天，交易累計 k 次，手上持有股票
			// 1. 第 i - 1 天，交易累計 k 次，持有但沒有賣出
			// 2. 第 i - 1 天，交易累計 k - 1 次，沒持有，現價買入 (k-1 < 0 代表首次交易，不用抓前值)
			if k-1 < 0 {
				dp[i][k][1] = max(dp[i-1][k][1], -prices[i])
			} else {
				dp[i][k][1] = max(dp[i-1][k][1], dp[i-1][k-1][0]-prices[i])
			}
		}
	}
	return dp[n-1][K-1][0]
}
// @lc code=end
